dev/archery/archery/linking.py

```diff
@@ -61,9 +64,71 @@ def list_dependency_names(self):
             names.append(name)
         return names
 
+    def _extract_symbols(self, symbol_info):
+        return [re.search(r'\S+$', line).group() for line in symbol_info if line]
+
+    def _remove_weak_symbols(self, lines):
+        return [line for line in lines if not re.search(r'\s[Ww]\s', line)]
+
+    def list_symbols_for_dependency(self, dependency, remove_symbol_versions=False):
+        result = _nm.run('-D', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        if remove_symbol_versions:
+            lines = [re.split('@@', line)[0] for line in lines]
+        return self._extract_symbols(lines)
+
+    def list_undefined_symbols_for_dependency(self, dependency,
+                                              remove_symbol_versions=False):
+        result = _nm.run('-u', dependency, stdout=subprocess.PIPE)
+        lines = result.stdout.decode('utf-8').splitlines()
+        if remove_symbol_versions:
+            lines = [re.split('@@', line)[0] for line in lines]
+        lines = self._remove_weak_symbols(lines)
+        return self._extract_symbols(lines)
+
+    def find_library_paths(self, libraries):
+        paths = {}
+        system = platform.system()
+        for lib in libraries:
+            paths[lib] = []
+            if system == 'Linux':
+                result = _ldconfig.run('-p', stdout=subprocess.PIPE)
+                lines = result.stdout.decode('utf-8').splitlines()
+                for line in lines:
+                    if lib in line:
+                        match = re.search(r' => (.*)', line)
+                        if match:
+                            paths[lib].append(match.group(1))
+            else:
+                raise ValueError(f"{platform} is not supported")
+        return paths
+
+
+def _check_undefined_symbols(dylib, allowed):
+    # Check for undefined symbols
+    undefined_symbols = dylib.list_undefined_symbols_for_dependency(dylib.path, True)
+    expected_lib_paths = dylib.find_library_paths(allowed)
+    all_paths = []
+
+    for paths in expected_lib_paths.values():
+        all_paths.extend(paths)
+
+    for lb_path in all_paths:
```

Comment:
Can we simplify this?

```suggestion
    expected_lib_paths = dylib.find_library_paths(allowed)
    for lb_path in expected_lib_paths.values():
```